### ISSUE TITLE
gi: "focus the first (or n-th) text input box on the page" does not work as before

### ISSUE THREAD

**Issue by ninjawisdom** [philc/vimium#4247]
If I type "gi", the cursor does not go to the end of the text form where some text has already been filled in, but instead goes to the head.

This issue has been occurring since around April 2023. I would like "gi" to behave as it did before.

*using Chrome ver.112.0.5615.121

**Comment by ninjawisdom** [comment]
Comment: This behavior seems to occur on Google, but not on other websites such as Amazon or Twitter. I apologize if it is not related to Vimium, but rather a Google-specific issue.

**Comment by bjzhush** [comment]
the same problem on my google page

**Comment by MarcoLucidi01** [comment]
hi, i've the same issue on firefox, i've investigated a bit since it's really
bothering me. it seems that `google.com` changed the html tag for the search bar
from `<input>` to `<textarea>` and the vimium function which triggers `.focus()`
on the selected "`gi` element" has a special case for `<textarea>` elements
which doesn't move the caret to the end of the text. this behaviour was
introduced with this commit [`a3a8ede00fa1405135d519dc0015e65953318a2c`](https://github.com/philc/vimium/pull/1943/commits/a3a8ede00fa1405135d519dc0015e65953318a2c)
with this motivation:

> Disable position-at-end in text areas.
>
> When we `simulateSelect` an input and the selection is at the start, we
> move it to the end.  This works well for single-line inputs.  However,
> the UX is *bad* for multiline inputs (such as text areas), and doubly so
> if the end of the input happens to be out of the viewport.
>
> This commit simply disables the repositioning of the selection within
> text areas.

now, *for my usage*, the google search bar acts as a "single-line input" 99% of
the times and the current behaviour breaks my usual workflow when i want to
quickly "refine" the search adding more terms at the end of the previous query.

the straightforward fix would be to just drop this `<textarea>` special case,
but i don't know if it's acceptable by people more seasoned in the project since
we would get back the "bad UX" mentioned in that commit message. maybe there is
a better solution. anyway i leave a quick patch here, if we agree that is ok, i
can make a PR:

``` diff
diff --git a/lib/dom_utils.js b/lib/dom_utils.js
index 102e93e4..9cb8197d 100644
--- a/lib/dom_utils.js
+++ b/lib/dom_utils.js
@@ -299,19 +299,17 @@ var DomUtils = {
       return handlerStack.bubbleEvent("click", {target: element});
     } else {
       element.focus();
-      if (element.tagName.toLowerCase() !== "textarea") {
-        // If the cursor is at the start of the (non-textarea) element's contents, send it to the end. Motivation:
-        // * the end is a more useful place to focus than the start,
-        // * this way preserves the last used position (except when it's at the beginning), so the user can
-        //   'resume where they left off'.
-        // NOTE(mrmr1993): Some elements throw an error when we try to access their selection properties, so
-        // wrap this with a try.
-        try {
-          if ((element.selectionStart === 0) && (element.selectionEnd === 0)) {
-            return element.setSelectionRange(element.value.length, element.value.length);
-          }
-        } catch (error) {}
-      }
+      // If the cursor is at the start of the element's contents, send it to the end. Motivation:
+      // * the end is a more useful place to focus than the start,
+      // * this way preserves the last used position (except when it's at the beginning), so the user can
+      //   'resume where they left off'.
+      // NOTE(mrmr1993): Some elements throw an error when we try to access their selection properties, so
+      // wrap this with a try.
+      try {
+        if ((element.selectionStart === 0) && (element.selectionEnd === 0)) {
+          return element.setSelectionRange(element.value.length, element.value.length);
+        }
+      } catch (error) {}
     }
   },
 

```

**Comment by zoriya** [comment]
I think we could conserve the textarea special behavior when there are multiple-lines in the textarea. To check that, we just need to check if \n are presents in the text.

**Comment by MarcoLucidi01** [comment]
yes, i think this could work, something like:

```
      if (element.tagName.toLowerCase() !== "textarea" || element.value.indexOf("\n") < 0) {
```

the only "problematic" case that comes to mind is when you have a single long
line that is wrapped into multiple lines in the `textarea`, in that case, the
text is displayed as multiple lines, but the string doesn't have `\n` chars,
but imho it's an acceptable tradeoff.

**Comment by philc** [comment]
I've just submitted this fix to the Chrome and Firefox stores as v1.67.7, and it should go live in the next day or two.
